compiler/rustc_lint/src/early.rs

```diff
@@ -247,6 +247,12 @@ impl<'a, T: EarlyLintPass> ast_visit::Visitor<'a> for EarlyContextAndPass<'a, T>
     }
 
     fn visit_where_predicate(&mut self, p: &'a ast::WherePredicate) {
+        use rustc_ast::{WhereBoundPredicate, WherePredicate};
+        if let WherePredicate::BoundPredicate(WhereBoundPredicate { bounded_ty, .. }) = p &&
+            let ast::TyKind::BareFn(b) = &bounded_ty.kind &&
+            b.generic_params.len() > 0 {
+                return;
```

Comment:
shouldn't this check for `ast::TyKInd::Paren` containing a `BareFn`

I think the condition should also include the check `bound_generic_params.is_empty()`.

THis should lint as `for<'b> for<'a> fn(Inv<'a>): Trait<'b>` is also correct:
```rust
#![deny(warnings)]

struct Inv<'a>(&'a mut &'a ());

trait Trait<'a> {}
impl<'b> Trait<'b> for for<'a> fn(Inv<'a>) {}


fn with_bound()
where
    for<'b> (for<'a> fn(Inv<'a>)): Trait<'b>, //~ ERROR unnecessary parentheses around type
{}

fn main() {
    with_bound();
}
```

Reply:
ah, but instead of changing the visiting behavior for all lints, you should instead add a method like `fn enter_lint_where_bound` or something to `EarlyLintPass` where the `UnusedParens` lint has special behavior. Similar to https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/trait.EarlyLintPass.html#method.enter_lint_attrs